doc/question.md

```diff
@@ -110,6 +88,81 @@ The `question.html` is regular HTML, with four special features:
+`prepare()` | | `correct_answers`, `params` | `options`, `variant_seed` | Final question preparation after element code has run. Can modify data as necessary. Return the modified `data` dictionary.
+`render()` | `html` (string) | | `correct_answers`, `editable`, `feedback`, `format_errors`, `options`, `panel`, `params`, `partial_scores`, `raw_submitted_answers`, `score`, `submitted_answers`, `variant_seed` | Render the HTML for one panel and return it as a string.
+`parse()` | | `format_errors`, `submitted_answers` | `correct_answers`, `options`, `params`, `raw_submitted_answers`, `variant_seed` | Parse the `data["submitted_answers"][var]` data entered by the student, modifying this variable. Return the modified `data` dictionary.
+`grade()` | | `correct_answers`, `feedback`, `format_errors`, `params`, `partial_scores`, `score`, `submitted_answers` | `options`, `raw_submitted_answers`, `variant_seed` | Grade `data["submitted_answers"][var]` to determine a score. Store the score and any feedback in `data["partial_scores"][var]["score"]` and `data["partial_scores"][var]["feedback"]`. Return the modified `data` dictionary.
+`file()` | `object` (string, byte-like, file-like) | | `correct_answers`, `filename`, `options`, `params`, `variant_seed` | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`. If `file()` returns nothing, an empty string will be used.
+
+A complete `question.html` and `server.py` example looks like:
+
+```html
+<!-- question.html -->
+
+<pl-question-panel>
+  <!-- params.x is defined by data["params"]["x"] in server.py's `generate()`. -->
+  <!-- params.operation defined by in data["params"]["operation"] in server.py's `generate()`. -->
+  If $x = {{params.x}}$ and $y$ is {{params.operation}} $x$, what is $y$?
+</pl-question-panel>
+
+<!-- y is defined by data["correct_answer"]["y"] in server.py's `generate()`. -->
+<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+```
+
+```python
+# server.py
+
+import random
+
+def generate(data):
+    # Generate random parameters for the question and store them in the data["params"] dict:
+    data["params"]["x"] = random.randint(5, 10)
+    data["params"]["operation"] = random.choice(["double", "triple"])
+
+    # Also compute the correct answer (if there is one) and store in the data["correct_answers"] dict:
+    if data["params"]["operation"] == "double":
+        data["correct_answers"]["y"] = 2 * data["params"]["x"]
+    else:
+        data["correct_answers"]["y"] = 3 * data["params"]["x"]
+
+def prepare(data):
+    # This function will run after all elements have run `generate()`.
+    # We can alter any of the element data here, but this is rarely needed.
+    pass
```

Comment:
I actually tested the skeleton this time, which caught several other errors as well (sorry, should've done this before). One of the errors was having an empty `prepare()`, so I added `pass`. Will this mess up the question in some invisible way (seems to work fine by my testing) if we override `prepare()` but make it empty?

Reply:
It's fine to have the functions present and not doing anything. We aren't actually overriding anything here. The flow is "run all the element functions", "do some internal bookkeeping", "run the server.py" function if it exists. Everything is designed to run correctly without a "server.py" existing at all, and if a function exists but passes then it's the same as if it just wasn't there.